### PR TITLE
New Dagger configuration

### LINKED ISSUE #2471: Standardize dependency injection across Collect
#### Problem description
So we are actively working on making more components within Collect loosely coupled so that we can increase the testability of the app as a whole and also the ability to test individual components. We could say that Dagger is one of the major proponents of this as it provides us with a DI solution that has been tested and proven by our industry to accelerate proper TDD workflows. Therefore it's usage throughout the app has been on an increase as we are constantly building out complex feature requirements that have several dependencies and scopes and this is the reason I think we should begin an official exercise of making all of this standard within the app and contribution guidelines.

- One of the first things we have to do get this initiative going is to do a comprehensive review of [SharedPreferences](https://github.com/opendatakit/collect/pull/1871) PR by @shobhitagarwal1612 as it's one of the most used components throughout the app and being to mock it would allow us to spin up test environments that require it a lot more easily. 

- Next on the list is the utilization of subcomponents and component dependencies. 
While doing a review of the current refactoring taking place on Collect HttpClient, I realized that [the refactor](https://github.com/opendatakit/collect/pull/2257) had a `HttpComponent` in it's midst. While utilizing that approach was a really good decision, there are some constraints associated with sub components that need to be addressed before it's introduction. Nonetheless, they are a good way to organize components and infrastructure along the lines of similar components, scopes and dependencies. Imagine a `ApplicationComponent` that contains all the general dependencies such as `SharedPreferences` and the `ApplicationContext` that all others components can inherit from, and all the  SMS related code in a`SmsComponent`. This approach could lead to easier dependency management, testing and component lifecyle. Before we can introduce this into the codebase we have to take a detailed look at which would work better, whether it be subcomponents or component dependencies, the learning curve and how which layers would be easier to try with with first as a proof of concept. 

- The flipside to this, is leaving subcomponents/component dependencies out of the picture and using the general `ApplicationComponent` approach that hosts all dependencies. This is the easier approach but it doesn't allow proper seperation of dependencies, scope and responsibilities. 

- Similar to what we did for permissions, we could make a list of the most important components or areas that need to become injectable and then we could have PRs to do this. While doing this evaluation we have to take into consideration the impact the changes will have on the codebase i.e the risk for regression and also verify that the area needs to be tested more thoroughly.

To begin this, I am asking @shobhitagarwal1612 to do a review of the best approaches we can take to accomplish this and if we would be able to mobilize enough resources to ensure that these changes would be contributor friendly! We have already started using most of these concepts throughout so doing this would be a great way to get everything formalized! 

Will update this with reading resources and more of my thoughts/research when I  finalize them.

### PR DESCRIPTION
**Pull request by seadowg**
Closes #2471.

This PRs attempts to rework the Dependency Injection approach in the codebase with an aim of using it to encourage and aid TDD workflows.

*Upfront disclaimer*: The approach taken in both the configuration and documentation is based on my opinion™ that the app should only use a DI framework when strictly necessary (I discuss this more below). Definitely happy for this PR to serve as a discussion around that approach!

I took the approach of starting with some documentation ([https://www.seadowg.com/dip-lesson/]()) around Dependency Inversion Principle that can hopefully serve as a guide to the kinds of code we're trying to achieve here and why it helps us write tests (ideally before writing code).

I then stripped back the Dagger (current DI framework in the app) configuration to be as simple and canonical as possible and added documentation to the configuration classes (mainly `AppDependencyComponent` and `AppDependencyModule`). This did mean removing some good work done as part of prep for an earlier refactor that was never finished (9b7cabc55fa4acae9b172f872ccc9f2d5e46f0b3). The configuration introduced here was slightly out of date given some newer Dagger changes. It also took a heavier approach on using Dagger which I think would be excellent for a team familiar with Dagger but I believe would cause problems for contributors new to Android and Dagger.

I then back-filled a test for an analytics event in the `InstanceUploaderListActivity` to drive out a Dagger setup that helped us write tests for black box objects (where we don't have access to constructors) such as Activity objects.

Then to hopefully drive out future use of the patterns in place I've added a section to the `CONTRIBUTING.md` documentation on the ways to inject and swap out dependencies.

#### What has been done to verify that this works as intended?

I ran automated tests and verified app appeared to behave normally.

#### Why is this the best possible solution? Were any other approaches considered?

I looked at some other DI frameworks (RoboGuice etc) but within the Java/Android world Dagger has become not only the most popular but also the most viable due to others lacking Android documentation/examples. 

From my own experience and from doing some research it does seem that there are (in my opinion much nicer and more intuitive) alternatives available within the Kotlin ecosystem such as Kodein and Koin. These rely on features within the Kotlin language however and wouldn't be very suitable unless we were looking to start using Kotlin heavily within the Collect codebase. This is a discussion for another time 😁.

Another alternative would have been to avoid using a DI framework. For Activities/Fragments/Services etc this can create inconsistent hacks around how dependencies are injected throughout the tests and the core code so although Dagger can seem a little heavyweight I do think it's worth having. 

Throughout the docs and within the configuration I have tried to push the idea of using Dagger only when it's really needed - if we can construct objects normally I think we should do just that. Hopefully this can reduce the need for people to understand Dagger and its configuration.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

There shouldn't be any changes to behaviour. There have been some changes to the core configuration of dependencies that are used across the app so it would be good to have a thorough sweep to verify the behaviour is all still intact.

#### Do we need any specific form for testing your changes? If so, please attach one.

No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
@seadowg I was just wondering if removing some of the boilerplate code with the android-dagger-support classes like `DaggerApplication`, `DaggerAppCompatActivity`, `DaggerFragment` would be a good idea or not. Did you not use them for a specific reason?

**Comment by seadowg** [comment]
Yes I think you're right! I think I missed this. As much as its covenient I think its better to not have frameworks like this integrated through inheritencie as it leads to heavy lock in/conflicts.

I'll fix this asap.

**Comment by seadowg** [comment]
@shobhitagarwal1612 While those classes cut out the injection line in a component they do (in my opinion) add more complexity to the Dagger setup (needing a DispatchingAndroidInjector etc). It also means that we have to have Dagger as part of our class hierarchy which can cause problems with Android or another framework introducing another base component we need to extend.

**Comment by zestyping** [comment]
Hi @seadowg, thanks for jumping in and tackling this.  I'm generally of a similar leaning on DI frameworks as you—I was heartened when you talked about "...[using] Dagger only when it's really needed - if we can construct objects normally I think we should do just that". :)

Okay, so I see a bunch of things happening in this PR.  I'll try to address them one by one.

- It cleans up the unused MVVM-related classes and GeoActivity.  Great.

- It takes some steps to simplify the use of Dagger, which is also good.

- It renames some `[foo]Component` classes to `[foo]DependencyComponent`.  I guess I don't have any particular feeling one way or another about that,  but I'm curious what motivated the renaming, and whether it reflects a convention or style that this PR intends to promote.

- It cleans up and adds helpful comments in a bunch of places.  Thanks for that!

- It makes the PermissionUtils constructor not depend on an Activity.  That makes sense.  I suspect that upon seeing how PermissionUtils takes an Activity as the first argument to every method, I might be immediately tempted to factor that out into a constructor method; maybe it would be good to add a brief comment explaining that the constructor deliberately avoids taking an Activity to avoid creating a dependency cycle.

- It makes InstanceUploaderListActivity get its PermissionUtils by injection instead of constructing one each time.  This makes sense, though I have a general question about injection below.

First let me say, overall, this PR is definitely a step forward as is and I'm in favour of merging.

It also creates an opportunity to ask two questions that I have about dependency injection:

1. Why are dependencies sometimes obtained via `@Inject`/`.inject(...)`, and sometimes obtained via `getApplicationComponent()` and calling a method on it directly?  How do we decide when to use which technique, or would it be better to pick one and stick with it?

2. (related to 1) Do we really need to use injection as much as we do?  Compare, for example, the code required to add a new dependency with injection:
```
    class AppDependencyModule {
        @Provider
        public Banana providesBanana() {
            return new Banana(...);
        }
    }

    interface AppDependencyComponent {
        void inject(BananaActivity activity);
    }

    class BananaActivity {
        @Inject
        Banana banana;

        public void onCreate(...) {
            ...
            getComponent().inject(this);
        }
    }
```
versus the code required to do the same thing without injection:
```
    class AppDependencyModule {
        public Banana getBanana() {
            return new Banana(...);
        }
    }

    class BananaActivity {
        Banana banana = DaggerUtils.getModule().getBanana();
    }
```
I think we can agree that the second is substantially simpler to write and to understand than the first. 
 Why (or when) is it better to do the first instead of the second?

**Comment by seadowg** [comment]
Thanks for the comments @zestyping! Really good questions. I'll give my thoughts below:

1. I'd generally advise using `@Inject` as it requires less noise in your code (you're not assigning to each field individually). Where I think it makes sense to use a helper method on the component to fetch an object is in test where you don't really want to go to the trouble of setting the test class up to be injected into. I'd look at `@Inject` as a sugar for the other really.

2. This is all down to Android life cycle and us not having access to a constructor. We're not in control of when the Activity (or Fragment, Service etc) is instantiated so we don't really want to assign fields until `onCreate` as our dependency might rely on some Context that doesn't exist yet. As an aside on this one: we never want to call the module @Provider directly as we'd bypass Dagger's scoping (`@Singleton`) etc. This is why when fetching individual instances (like in question 1) we add a helper to the component.

On the bullet points:

* The rename is an attempt to make it more obvious to newcomers. You can call these classes anything you want so I was hoping adding the extra 'Dependency' would help people that hadn't been exposed to Dagger before.
* In terms of the PermissionUtils refactor. Happy to add a comment to the code.

Let me know if you have any more questions!

**Comment by yanokwa** [comment]
@seadowg Once the conflicts are resolved, I think we should be ready to merge.

**Comment by zestyping** [comment]
Thanks, @seadowg.

Yes, I agree that `@Inject` is syntactic sugar for an assignment statement.  My claim is that it is more opaque and more complex than an assignment statement.  Even if the dependency needs to be constructed at a later point in the Activity lifecycle, like so:

```
    class AppDependencyModule {
        public Banana getBanana() {
            return new Banana(...);
        }
    }

    class BananaActivity {
        private Banana banana;

        public void onCreate(...) {
            ...
            banana = DaggerUtils.getHelper().getBanana();
        }
    }
```

...this code is still substantially simpler and easier to understand than code that uses `@Inject`.

The drawbacks of `@Inject` are many:

  * Using `@Inject` requires the involvement of a third file, the AppDependencyComponent interface, which we must remember to edit even though it is mentioned nowhere in the code.
  * Using `@Inject` couples the code to distant, complicated machinery that has nothing to do with the task the code is trying to achieve.
  * Using `@Inject` obscures the assignment to the instance member (the assignment to `banana` cannot be seen anywhere in the file), whereas with a regular assignment it is obvious when and how the instance member is initialized.
  * Using `@Inject` violates the guarantees of access modifiers.  For example, the defining guarantee of `protected` is that we can determine all the possible values that a protected field can have simply by looking for any assignment to the field in the class or its subclasses; `@Inject` breaches that basic contract.
  * Using `@Inject` prevents the use of `private` or `final` on members that should properly be declared private or final.
  * All the members marked with `@Inject` must be initialized at the same time.

If we run into one of these limitations, like wanting to declare a member `private` or initialize it at a different time, then we must either awkwardly accommodate `@Inject`'s limitations or awkwardly use a mixture of `@Inject` annotations or assignments, forcing the writer to make extra decisions and the reader to bear the burden of understanding them.

That's a lot of mental overhead for both writers and readers to carry around.  For such heavy costs, a feature would have to provide some major upsides to be a net benefit.  But the "sugar" that `@Inject` provides is modest: it allows us to replace N lines of code (N explicit assignments) with N+1 shorter lines of code (N `@Inject` annotations plus one call to `.inject()`).  That's it, as far as I can tell.

That's why I'm skeptical that `@Inject` is worthwhile.  Again, let me be clear that I don't intend to block this pull request; it should proceed.  I'd just like us to discuss the possibility of avoiding `@Inject` in the longer term, in the interest of a more readable, maintainable codebase.

**Comment by seadowg** [comment]
I've rebased master to deal with conflicts. 

Definitely interested in a discussion @zestyping. I'll post some thoughts as soon as I get a chance.

**Comment by seadowg** [comment]
Finally got a chance to reply to your thoughts @zestyping. Sorry about the delay! I have a few general points on the topic of "to `@Inject` or not to `@Inject`" - I'll mostly present the positive case as I think you've made great argmumnets around the problems.

I'll try to first comment on each one of your bullets however:

```md
Using @Inject requires the involvement of a third file, the AppDependencyComponent interface, which we must remember to edit even though it is mentioned nowhere in the code.
```

Definitely something I dislike about Dagger. The module and component setup is a consequence of Dagger carrying out injection at compile time which while efficient adds a lot of configuration and also makes it (in my opinion) quite hard to learn and understand. I mainly proceeded down the Dagger route due to its popularity within Android. For further explorations down this road we could certainly weigh up the pros and cons of a potentially simpler solution such as `Guice` (which has more self contained feeling setup).

```md
Using @Inject couples the code to distant, complicated machinery that has nothing to do with the task the code is trying to achieve.
```

Potentially. A lot of the goal of a DI framework is to intentionally **decouple** the class from the machinery that injects into it. In my mind this creates a world more similar to that of a constructor where we can declare what we need but don't need to specify where it comes from. Its hard to go all the way with Android however as we generally need some kind of 'inject' operation in `onCreate` that gives the game away anyway 😄.

```md
Using @Inject obscures the assignment to the instance member (the assignment to banana cannot be seen anywhere in the file), whereas with a regular assignment it is obvious when and how the instance member is initialized.
```

I think this relates to the above point.

```md
Using @Inject violates the guarantees of access modifiers. For example, the defining guarantee of protected is that we can determine all the possible values that a protected field can have simply by looking for any assignment to the field in the class or its subclasses; @Inject breaches that basic contract.
```

Absolutely a tradeoff of a lot of these frameworks/containers. As an aside, we'res seeing some nicer solutions within Kotlin out of interest that are able to maintain the purity in this regard (see [Kodein](https://github.com/Kodein-Framework/Kodein-DI) as an example).

```md
Using @Inject prevents the use of private or final on members that should properly be declared private or final.
```

For Dagger certainly we sacrifice `private`. Its hard for us to have a solution for this that's ever `final` however as we'll always be using field declaration (as assigning fields at `onCreate` is already to late for `final`).

```md
All the members marked with @Inject must be initialized at the same time.
```

Personally I think this is an advantage. I'd want to think of things we are `@Inject`ing as "dependencies" for an object and so would be required as part of that object's construction. As an aside, we generally want to avoid any state within our Activities due to the complexities we can run into with lifecycle so I'd argue we only want fields for objects we are treating as these upfront "dependencies".

---

I can totally understand that if people aren't used to seeing `@Inject` or similar within a codebase then it's going to lead to some confusion and the need for some level of documentation. It is just a tool so if the majority of people contributing are being hindered rather than helped I could see how it would make sense to drop it. Generally the advantage of keeping `@Inject` to me feels like it comes down to it being a "standard" of sorts:

* This pattern is pretty common in my experience in the JVM (and friends) world (both Android and server with the likes of Spring). `@Inject` is actually part of the Java standard lib JSX package and was [standardised on in 2009](http://javax-inject.github.io/javax-inject/).
* Dropping `@Inject` probably removes a lot of the reasons to have a DI framework as we'd be using a well known framework in an undocumented/supported way. This could lead to problems for both people who know Dagger and also people using its docs to understand what we're up to.
* Google mention approaches to managing dependencies in their
[architecture docs](https://developer.android.com/jetpack/docs/guide#manage-dependencies) and Dagger is their suggested implementation. They do also talk about using a service locator pattern which we could just implement ourselves. My only worry on this would be the extra maintenance and not having the advantage of another team thinking about bugs, memory leaks etc.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -95,19 +95,92 @@ In addition to contributing code, you can help to triage issues. This can includ
...
+
 ## Strings
 Always use [string resources](https://developer.android.com/guide/topics/resources/string-resource.html) instead of literal strings. This ensures wording consistency across the project and also enables full translation of the app. Only make changes to the base `res/values/strings.xml` English file and not to the other language files. The translated files are generated from [Transifex](https://www.transifex.com/opendatakit/collect/) where translations can be submitted by the community. Names of software packages or other untranslatable strings should be placed in `res/values/untranslated.xml`.
 
+## Dependency injection
+
+As much as possible to facilitate simpler, more modular and more testable components you should follow the Dependency Inversion principle in Collect Code. An example tutorial on this concept can be found [here](https://www.seadowg.com/dip-lesson/).
+
+Because many Android components (Activity and Fragment for instance) don't allow us control over their constructors Collect uses [Dagger](https://google.github.io/dagger/) to 'inject' dependencies. The configuration for Dagger can be found in `AppDepdendencyComponent`.
```
Comment (cooperka) [inline CONTRIBUTING.md]: Let's link to the actual `AppDepdendencyComponent` source

File: collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java [hunk 2]
```diff
@@ -67,12 +65,14 @@
...
 import static org.odk.collect.android.preferences.GeneralKeys.KEY_SMS_PREFERENCE;
-//import static org.odk.collect.android.preferences.GeneralKeys.KEY_SUBMISSION_TRANSPORT_TYPE;
 import static org.odk.collect.android.preferences.GeneralKeys.KEY_SUBMISSION_URL;
-import static org.odk.collect.android.utilities.DialogUtils.showDialog;
 import static org.odk.collect.android.preferences.GeneralKeys.KEY_TRANSPORT_PREFERENCE;
+import static org.odk.collect.android.utilities.DialogUtils.showDialog;
 import static org.odk.collect.android.utilities.gdrive.GoogleAccountsManager.REQUEST_ACCOUNT_PICKER;
 
+//import android.preference.ListPreference;
+//import static org.odk.collect.android.preferences.GeneralKeys.KEY_SUBMISSION_TRANSPORT_TYPE;
```
Comment (cooperka) [inline collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: I assume these were auto-updated, but please remove :)

**Review by cooperka — APPROVED** [review]
Thank you both for the detailed discussion, it gave me a lot of context. There's a lot of good cleanup here and I think it's absolutely worth merging as-is.

However I agree with Ping that the `@Inject` annotation is more opaque and less maintainable than a plain old assignment in `onCreate`. The main benefit seems to be the ability to use `@Singleton` but that could be addressed with a simple static variable to store the singleton and always return it.

I've worked on a large Android app in the past that heavily used DI, and I experienced nearly every one of Ping's bulleted frustrations. In particular I think the "magic" obfuscation is painful to newcomers, and being open-source that's a big factor. "Where is the var coming from? When is it initialized? How do I change it? How do I use it in another file?" All of these questions are immediately answered when assigning rather than `@Inject`ing, while still preserving the testability benefits of DI.

Again, I think we should merge this. Whether or not to `@Inject` is a side-question that can be changed later with relative ease. Shall we set some sort of date to come back and decide, after developers have a chance to use the new logic? Any other thoughts, @zestyping?

File: CONTRIBUTING.md (hunk 1 above)
Comment (seadowg) [reply CONTRIBUTING.md]: Ah nice idea will do

**Comment by seadowg** [comment]
@cooperka yeah I think it's good to punt on "To `@Inject` or not to `@Inject`" right now. If people are happy pulling it out later then it might just make sense to pull out Dagger entirely. My impression from the apps I've worked on is unless you're using all the power features (fragment, activity scopes etc) then you're dealing with some confusing configuration for not a lot of extra benefit. Would be interesting to play with simpler frameworks or just build in a (probably very) simple service locator ourselves (which I've done before).

P.S.  I just realized my previous post's link tot he Google Architecture docs was hilariously just a link back to this PR so I've corrected that 🙈

File: collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java (hunk 2 above)
Comment (cooperka) [inline collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: (I think you removed different ones than these)

**Comment by cooperka** [comment]
@yanokwa do you think we need more eyes before labeling for testing?

File: collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java (hunk 2 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: You are totally correct. My bad. Will fix now - I'll leave the commented out fields deleted as well.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: Actually it looks like the commented out lines have something to do with the recent disabling of SMS features (https://github.com/opendatakit/collect/commit/d47e3312ae956f7af421505b44e665d5a181e925#diff-28149f9be7f929447487b6010fea53c3). 

Definitely happy to just delete the commented code (its in version control we can get it back) but maybe we should leave it alone in this PR? I can revert the delete of the comment fields if so.

Comment (cooperka) [reply collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: I see. Yeah I guess that should be reverted then. Thanks for clarifying. Hopefully all that code can just be removed soon.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/preferences/ServerPreferencesFragment.java:74]: @cooperka reverted!

**Comment by cooperka** [comment]
@opendatakit-bot label "needs testing"

**Comment by yanokwa** [comment]
@mmarciniak90 When you test this, I think a basic end-to-end (get all widgets, fill it out, save, edit, submit) test with a fresh install should be good enough.

**Comment by kkrawczyk123** [comment]
**Tested with success!!**
Verified on Androids 4.2, 4.4, 5.1, 6.0, 7.1 and 8.1.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

**Review by shobhitagarwal1612 — APPROVED** [review]

**Pull request merged**